python/treelite/gtil/gtil.py

```diff
@@ -26,34 +60,74 @@ def predict(model: Model, data: np.ndarray, nthread: int = -1, pred_margin: bool
         2D NumPy array, with which to run prediction
     nthread : :py:class:`int <python:int>`, optional
         Number of CPU cores to use in prediction. If <= 0, use all CPU cores.
-    pred_margin : :py:class:`bool <python:bool>`, optional
-        Whether to produce raw margin scores
+    pred_margin : bool, optional
+        Deprecated. Set predict_type="raw" instead.
+    predict_type : string, optional
+        One of the following:
+        * "default": Usual prediction method. Sum over trees and apply post-processing.
+        * "raw": Sum over trees, but don't apply post-processing; get raw margin scores
+                 instead.
+        * "leaf_id": Output one (integer) leaf ID per tree.
+        * "score_per_tree": Output one or more margin scores per tree.
```

Comment:
Without having reviewed the associated implementation yet, would it in principle be possible to obtain the raw, leaf_id, and score_per_tree result types after completing the prediction? In that case we could simply return an object that can be used to obtain all of this data. Something like this:

```python
y = predict(model, X)  # standard approach

# Obtain alternative (rich) result types:
results = predict_results(model, X)
results.prediction
results.raw
results.leaf_id
# and so on ...
```

Using a separate method instead of changing the return type based on a function argument also seems like a slightly better design to me in that case.

The standard predict method is then basically just short for: `y = predict_results(model, X).prediction`.

Reply:
No, each mode of prediction has sufficiently different implementation.